ClickUp integration docs (#878)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/clickup.mdx

</div>

## Getting started

Before configuring ClickUp as a destination in RudderStack, verify if the source platform is supported by ClickUp by referring to the table below:

| Connection Mode | Web | Mobile | Server |
| :--- | :--- | :--- | :--- |
| Device mode | - | - | - |
| Cloud mode | Supported  | Supported | Supported |

<div class="infoBlock">
To know more about the difference between cloud mode and device mode in RudderStack, refer to the <Link to="/destinations/rudderstack-connection-modes/">RudderStack Connection Modes</Link> guide.
</div>

Once you have confirmed that the source platform supports sending events to ClickUp, follow these steps:

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add a source. Then, from the list of destinations, select **ClickUp**.
2. Assign a name to the destination and click **Continue**.

## Connection settings

To successfully configure ClickUp as a destination, you will need to configure the following settings:

<img src="../../assets/event-stream-destinations/clickup-connection-settings-1.png" alt="ClickUp connection settings" />

- **API Token**: Enter your ClickUp API token.

> <div class="infoBlock">

Reviewer: Can remove here and add it in the below note (next suggestion)

Author: We're referencing the individual FAQ in each info block. In that case we will need a generic `<Link to="#faq">FAQ</Link>` tag. I'll make the change.

For more information on obtaining your ClickUp API token, refer to the <Link to="#where-can-i-find-the-clickup-api-token">FAQ</Link> section below.
</div>

- **List ID**: Enter your ClickUp list ID.

<div class="infoBlock">

> For more information on obtaining your ClickUp list ID, refer to the <Link to="#where-can-i-find-the-clickup-list-id">FAQ</Link> section below.

Reviewer: ```suggestion
For more information on obtaining your ClickUp API token and list ID, refer to the <Link to="#where-can-i-find-the-clickup-list-id">FAQ</Link> section below.
```

Author: Need to change the link tag here as discussed above.

</div>

- **Mapping to add custom fields while creating a task**: Use this setting to map a specific RudderStack event property to a ClickUp Custom Field.

<div class="infoBlock">
For more information on using this setting, refer to the <Link to="#adding-custom-fields">Adding custom fields</Link> section below.
</div>

<div class="infoBlock">
You can provide multiple ClickUp custom fields associated with different lists. However, note that the custom field name <strong>must be unique</strong> within a list to successfully create a task.
</div>

> - **Client-side Events Filtering**: This setting lets you specify which events should be allowed to flow through to ConvertFlow.

Reviewer: ```suggestion
- **Client-side Events Filtering**: This setting lets you specify which events should be allowed to flow through to ClickUp.
```

<img src="../../assets/event-stream-destinations/clickup-connection-settings-2.png" alt="ClickUp connection settings" />

> <div class="warningBlock">

Reviewer: Not sure if we should link it to "Client-side Events Filtering" guide as that guide is for device mode supported destinations and ClickUp supports only cloud mode.

Author: @ishachopra0401 I was initially conflicted with the same thought, but thought of adding it anyway. we can remove this.

<img src="../../assets/event-stream-destinations/clickup-custom-field.png" alt="ClickUp custom field dashboard setting" />

You can specify the **ClickUp Custom Field Name** corresponding to the <Link to="#how-can-i-create-custom-fields-in-clickup">custom field you created in the ClickUp dashboard</Link> and the **RudderStack Payload Property** corresponding to the event property present in the payload.

RudderStack uses the custom field name to retrieve the `id` using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint.

### Supported custom fields

RudderStack supports specifying the following ClickUp custom fields in the dashboard settings:

| Custom Field | Notes |
| :-----| :-----| 
| URL | Must be in a valid URL format. |
| Email | Must be in a valid email format. | 
| Phone | Must be a valid phone number with a country code, for example, `+1 123 456 7890`. |
| Date | Must be a valid date in a datetime format, for example, `2022-02-25T13:39:21.032Z`. |
| Text | Must be a string. |
| Checkbox | Must be in a Boolean format. |
| Number | - |

> | Currency | Must be a number |

Reviewer: ```suggestion
| Currency | Must be a number. |
```

| Emoji (integer) | Used for rating. The value must be in a range defined in the ClickUp dashboard. |
| Location | Latitude/longitude and the formatted address must be passed as the [Google Geocoding API](https://developers.google.com/maps/documentation/geocoding/overview?hl=en). |

> | Dropdown | RudderStack will use the dropdown option name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |

Reviewer: ```suggestion
| Dropdown | RudderStack uses the dropdown option name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |
```

> | Dropdown | RudderStack will use the dropdown option name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |

Reviewer: ```suggestion
| Dropdown | RudderStack uses the dropdown option name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |
```

> | Label | RudderStack will the label name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |

Reviewer 2: Typo here. `use` is missing.
 RudderStack will `use` the label ...

> | Label | RudderStack will the label name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |

Author: ```suggestion
| Label | RudderStack will use the label name to find the UUID using the [`field`](https://api.clickup.com/api/v2/list/list_id/field) endpoint. |
```

| `properties.startDate` | `start_date` | Datetime |
| `properties.includeStartDateTime` | `start_date_time` | Boolean |
| `properties.notifyAll` | `notify_all` | Boolean | 
| `properties` <br />(Only fields mentioned in the dashboard mapping) | `custom_fields` | Array | 

## Rate limits

Note that the ClickUp enforces rate limits per OAuth and personal token. This rate limit depends on your ClickUp plan and is summarized in the following table:

| ClickUp plan | Requests allowed per minute |
| :----- | :------ | 
| Free Forever, Unlimited, Business | 100 |
| Business Plus | 1000 | 
| Enterprise | 10000 | 

<div class="infoBlock">
For more information, refer to the <a href="https://clickup.com/pricing">ClickUp Pricing</a> page.
</div>

## FAQ

### Where can I find the ClickUp API Token?

To obtain your ClickUp API token, follow these steps:

1. Log into your [ClickUp dashboard](https://app.clickup.com/).
2. From the left sidebar, click the **Settings** button at the bottom and go to **MY APPS** > **Apps**.
3. You can find the API token listed here, as shown:

<img src="../../assets/event-stream-destinations/clickup-api-token.png" alt="ClickUp API token" />

<div class="infoBlock">

> To create a new app with the ClickUp API, refer to this <a href="https://help.clickup.com/hc/en-us/articles/6303422883095-Create-your-own-app-with-the-ClickUp-API">ClickUp Help Guide</a>.

Reviewer 2: We can remove this. As we are not using the `OAuth` for authentication. 
Currently, We are only supporting authentication with `API token`.

Author: This note is now removed.
